File: 1504. Count Submatrices With All Ones/Solution.cpp
#include <deque>
#include <stack>
#include <vector>
using namespace std;

class Solution {
public:
    int numSubmat(vector<vector<int>>& mat)
    {
        int m = mat.size(), n = mat[0].size(), ans = 0;
        vector<int> height(n, 0);
        for (int i = 0; i < m; ++i) {
            stack<int> mono_stk(deque{-1});
            for (int j = 0; j < n; ++j) {
                height[j] = mat[i][j] ? height[j] + 1 : 0;
                while (mono_stk.size() > 1 && height[j] <= height[mono_stk.top()])
                    mono_stk.pop();
                int last = mono_stk.top();
                mono_stk.emplace(j);
                mat[i][j] = height[j] * (j - last);
                if (last != -1)
                    mat[i][j] += mat[i][last];
                ans += mat[i][j];
            }
        }
        return ans;
    }
};

File: 1504. Count Submatrices With All Ones/Solution.py
class Solution:
    def numSubmat(self, mat: list[list[int]]) -> int:
        m, n = len(mat), len(mat[0])
        height = [0] * n
        ans = 0
        for i in range(m):
            mono_stack = [-1]
            for j in range(n):
                height[j] = height[j] + 1 if mat[i][j] == 1 else 0
                while len(mono_stack) > 1 and height[j] <= height[mono_stack[-1]]:
                    mono_stack.pop()
                last = mono_stack[-1]
                mono_stack.append(j)
                mat[i][j] = height[j] * (j - last)
                if last != -1:
                    mat[i][j] += mat[i][last]
                ans += mat[i][j]
        return ans
